File: mns-core/src/test/java/cz/fb/manaus/core/model/BetTest.java
package cz.fb.manaus.core.model;

import com.fasterxml.jackson.databind.ObjectMapper;
import cz.fb.manaus.core.provider.ExchangeProvider;
import cz.fb.manaus.core.test.AbstractLocalTestCase;
import cz.fb.manaus.core.test.CoreTestFactory;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Date;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

public class BetTest extends AbstractLocalTestCase {

    @Autowired
    private ExchangeProvider provider;

    @Test
    public void testHalfMatched() {
        assertTrue(createBet(provider.getMinAmount()).isHalfMatched());
        assertTrue(createBet(1.5).isHalfMatched());
        assertFalse(createBet(0d).isHalfMatched());
        assertFalse(createBet(0.8d).isHalfMatched());
    }

    @Test
    public void testMatched() {
        assertTrue(createBet(provider.getMinAmount()).isMatched());
        assertTrue(createBet(1.5).isMatched());
        assertFalse(createBet(0d).isMatched());
        assertTrue(createBet(0.8d).isMatched());
    }

    @Test
    public void testSerialization() throws Exception {
        var mapper = new ObjectMapper();
        var original = new Bet("111", "222", 333,
                new Price(3d, 2d, Side.BACK), new Date(), 0d);

        var serialized = mapper.writer().writeValueAsString(original);
        Bet restored = mapper.readerFor(Bet.class).readValue(serialized);
        assertThat(restored.getRequestedPrice(), is(original.getRequestedPrice()));
        assertThat(restored.getPlacedDate(), is(original.getPlacedDate()));
        var doubleSerialized = mapper.writer().writeValueAsString(restored);
        assertEquals(serialized, doubleSerialized);
    }

    private Bet createBet(double matchedAmount) {
        var marketId = CoreTestFactory.MARKET_ID;
        var selectionId = CoreTestFactory.DRAW;
        var requestedPrice = new Price(3d, provider.getMinAmount(), Side.LAY);
        var date = Instant.now().minus(2, ChronoUnit.HOURS);
        return new Bet("1", marketId, selectionId, requestedPrice, Date.from(date), matchedAmount);
    }
}

File: mns-core/src/test/java/cz/fb/manaus/core/model/BetTest.kt
package cz.fb.manaus.core.model

import com.fasterxml.jackson.databind.ObjectMapper
import cz.fb.manaus.core.test.AbstractLocalTestCase
import org.junit.Assert.*
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import java.time.Instant
import java.time.temporal.ChronoUnit

val betTemplate = Bet(
    betId = "111",
    marketId = "222",
    selectionId = SEL_HOME,
    requestedPrice = Price(3.0, 2.0, Side.BACK),
    placedDate = Instant.now()
)


class BetTest : AbstractLocalTestCase() {

    @Autowired
    private lateinit var mapper: ObjectMapper

    @Test
    fun `half matched`() {
        assertTrue(createBet(mbProvider.minAmount).isHalfMatched)
        assertTrue(createBet(1.5).isHalfMatched)
        assertFalse(createBet(0.0).isHalfMatched)
        assertFalse(createBet(0.8).isHalfMatched)
    }

    @Test
    fun `fully matched`() {
        assertTrue(createBet(mbProvider.minAmount).isMatched)
        assertTrue(createBet(1.5).isMatched)
        assertFalse(createBet(0.0).isMatched)
        assertTrue(createBet(0.8).isMatched)
    }

    @Test
    fun `json marshall`() {

        val serialized = mapper.writer().writeValueAsString(betTemplate)
        val restored = mapper.readerFor(Bet::class.java).readValue<Bet>(serialized)
        assertEquals(betTemplate.requestedPrice, restored.requestedPrice)
        assertEquals(betTemplate.placedDate, restored.placedDate)
        val doubleSerialized = mapper.writer().writeValueAsString(restored)
        assertEquals(serialized, doubleSerialized)
    }

    private fun createBet(matchedAmount: Double): Bet {
        val requestedPrice = Price(3.0, mbProvider.minAmount, Side.LAY)
        val date = Instant.now().minus(2, ChronoUnit.HOURS)
        return Bet(
            betId = "1",
            marketId = market.id,
            selectionId = 1000L,
            requestedPrice = requestedPrice,
            placedDate = date,
            matchedAmount = matchedAmount
        )
    }
}